diff --git a/lld/ELF/MarkLive.cpp b/lld/ELF/MarkLive.cpp
index c43022a6b9d3..6430afcac4f2 100644
--- a/lld/ELF/MarkLive.cpp
+++ b/lld/ELF/MarkLive.cpp
@@ -53,21 +53,22 @@ struct LiveReason {
   StringRef desc;
 };
 
-template <class ELFT, bool TrackWhyLive> class MarkLive {
+template <bool TrackWhyLive> class MarkLive {
 public:
   MarkLive(Ctx &ctx, unsigned partition) : ctx(ctx), partition(partition) {}
 
-  void run();
+  LLVM_ATTRIBUTE_NOINLINE void seed();
+  LLVM_ATTRIBUTE_NOINLINE void reportWhyLive() const;
   void moveToMain();
   void printWhyLive(Symbol *s) const;
+  template <class ELFT> void mark();
 
 private:
   void enqueue(InputSectionBase *sec, uint64_t offset, Symbol *sym,
                LiveReason reason);
   void markSymbol(Symbol *sym, StringRef reason);
-  void mark();
 
-  template <class RelTy>
+  template <class ELFT, class RelTy>
   void resolveReloc(InputSectionBase &sec, const RelTy &rel, bool fromFDE);
 
   void scanEhFrameSection(EhInputSection &eh);
@@ -111,11 +112,10 @@ static uint64_t getAddend(Ctx &, InputSectionBase &sec,
   return rel.r_addend;
 }
 
-template <class ELFT, bool TrackWhyLive>
-template <class RelTy>
-void MarkLive<ELFT, TrackWhyLive>::resolveReloc(InputSectionBase &sec,
-                                                const RelTy &rel,
-                                                bool fromFDE) {
+template <bool TrackWhyLive>
+template <class ELFT, class RelTy>
+void MarkLive<TrackWhyLive>::resolveReloc(InputSectionBase &sec,
+                                          const RelTy &rel, bool fromFDE) {
   // If a symbol is referenced in a live section, it is used.
   Symbol *sym;
   if constexpr (std::is_same_v<RelTy, Relocation>) {
@@ -200,15 +200,15 @@ void MarkLive<ELFT, TrackWhyLive>::resolveReloc(InputSectionBase &sec,
 // A possible improvement would be to fully process .eh_frame in the middle of
 // the gc pass. With that we would be able to also gc some sections holding
 // LSDAs and personality functions if we found that they were unused.
-template <class ELFT, bool TrackWhyLive>
-void MarkLive<ELFT, TrackWhyLive>::scanEhFrameSection(EhInputSection &eh) {
+template <bool TrackWhyLive>
+void MarkLive<TrackWhyLive>::scanEhFrameSection(EhInputSection &eh) {
   if (TrackWhyLive)
     whyLive.try_emplace(&eh,
                         LiveReason{std::nullopt, "exception handling frame"});
   ArrayRef<Relocation> rels = eh.rels;
   for (const EhSectionPiece &cie : eh.cies)
     if (cie.firstRelocation != unsigned(-1))
-      resolveReloc(eh, rels[cie.firstRelocation], false);
+      resolveReloc<void>(eh, rels[cie.firstRelocation], false);
   for (const EhSectionPiece &fde : eh.fdes) {
     size_t firstRelI = fde.firstRelocation;
     if (firstRelI == (unsigned)-1)
@@ -216,7 +216,7 @@ void MarkLive<ELFT, TrackWhyLive>::scanEhFrameSection(EhInputSection &eh) {
     uint64_t pieceEnd = fde.inputOff + fde.size;
     for (size_t j = firstRelI, end2 = rels.size();
          j < end2 && rels[j].offset < pieceEnd; ++j)
-      resolveReloc(eh, rels[j], true);
+      resolveReloc<void>(eh, rels[j], true);
   }
 }
 
@@ -242,10 +242,9 @@ static bool isReserved(InputSectionBase *sec) {
   }
 }
 
-template <class ELFT, bool TrackWhyLive>
-void MarkLive<ELFT, TrackWhyLive>::enqueue(InputSectionBase *sec,
-                                           uint64_t offset, Symbol *sym,
-                                           LiveReason reason) {
+template <bool TrackWhyLive>
+void MarkLive<TrackWhyLive>::enqueue(InputSectionBase *sec, uint64_t offset,
+                                     Symbol *sym, LiveReason reason) {
   // Usually, a whole section is marked as live or dead, but in mergeable
   // (splittable) sections, each piece of data has independent liveness bit.
   // So we explicitly tell it which offset is in use.
@@ -277,8 +276,8 @@ void MarkLive<ELFT, TrackWhyLive>::enqueue(InputSectionBase *sec,
 }
 
 // Print the stack of reasons that the given symbol is live.
-template <class ELFT, bool TrackWhyLive>
-void MarkLive<ELFT, TrackWhyLive>::printWhyLive(Symbol *s) const {
+template <bool TrackWhyLive>
+void MarkLive<TrackWhyLive>::printWhyLive(Symbol *s) const {
   // Skip dead symbols. A symbol is dead if it belongs to a dead section.
   if (auto *d = dyn_cast<Defined>(s)) {
     auto *sec = dyn_cast_or_null<InputSectionBase>(d->section);
@@ -335,8 +334,8 @@ void MarkLive<ELFT, TrackWhyLive>::printWhyLive(Symbol *s) const {
   }
 }
 
-template <class ELFT, bool TrackWhyLive>
-void MarkLive<ELFT, TrackWhyLive>::markSymbol(Symbol *sym, StringRef reason) {
+template <bool TrackWhyLive>
+void MarkLive<TrackWhyLive>::markSymbol(Symbol *sym, StringRef reason) {
   if (auto *d = dyn_cast_or_null<Defined>(sym))
     if (auto *isec = dyn_cast_or_null<InputSectionBase>(d->section))
       enqueue(isec, d->value, sym, {std::nullopt, reason});
@@ -345,8 +344,7 @@ void MarkLive<ELFT, TrackWhyLive>::markSymbol(Symbol *sym, StringRef reason) {
 // This is the main function of the garbage collector.
 // Starting from GC-root sections, this function visits all reachable
 // sections to set their "Live" bits.
-template <class ELFT, bool TrackWhyLive>
-void MarkLive<ELFT, TrackWhyLive>::run() {
+template <bool TrackWhyLive> void MarkLive<TrackWhyLive>::seed() {
   // Add GC root symbols.
 
   // Preserve externally-visible symbols if the symbols defined by this
@@ -356,10 +354,8 @@ void MarkLive<ELFT, TrackWhyLive>::run() {
       markSymbol(sym, "externally visible symbol; may interpose");
 
   // If this isn't the main partition, that's all that we need to preserve.
-  if (partition != 1) {
-    mark();
+  if (partition != 1)
     return;
-  }
 
   markSymbol(ctx.symtab->find(ctx.arg.entry), "entry point");
   markSymbol(ctx.symtab->find(ctx.arg.init), "initializer function");
@@ -433,43 +429,43 @@ void MarkLive<ELFT, TrackWhyLive>::run() {
       cNamedSections[ctx.saver.save("__stop_" + sec->name)].push_back(sec);
     }
   }
+}
 
-  mark();
-
-  if (TrackWhyLive) {
-    const auto handleSym = [&](Symbol *sym) {
-      if (llvm::any_of(ctx.arg.whyLive, [sym](const llvm::GlobPattern &pat) {
-            return pat.match(sym->getName());
-          }))
-        printWhyLive(sym);
-    };
+template <bool TrackWhyLive>
+void MarkLive<TrackWhyLive>::reportWhyLive() const {
+  const auto handleSym = [&](Symbol *sym) {
+    if (llvm::any_of(ctx.arg.whyLive, [sym](const llvm::GlobPattern &pat) {
+          return pat.match(sym->getName());
+        }))
+      printWhyLive(sym);
+  };
 
-    for (Symbol *sym : ctx.symtab->getSymbols())
-      handleSym(sym);
-    // Handle local symbols, skipping the symbol at index 0 and section
-    // symbols, which usually have empty names and technically not live. Note:
-    // a live section may lack an associated section symbol, making them
-    // unreliable liveness indicators.
-    for (ELFFileBase *file : ctx.objectFiles)
-      for (Symbol *sym : file->getSymbols())
-        if (sym->isLocal() && sym->isDefined() && !sym->isSection())
-          handleSym(sym);
-  }
+  for (Symbol *sym : ctx.symtab->getSymbols())
+    handleSym(sym);
+  // Handle local symbols, skipping the symbol at index 0 and section symbols,
+  // which usually have empty names and technically not live. Note: a live
+  // section may lack an associated section symbol, making them unreliable
+  // liveness indicators.
+  for (ELFFileBase *file : ctx.objectFiles)
+    for (Symbol *sym : file->getSymbols())
+      if (sym->isLocal() && sym->isDefined() && !sym->isSection())
+        handleSym(sym);
 }
 
-template <class ELFT, bool TrackWhyLive>
-void MarkLive<ELFT, TrackWhyLive>::mark() {
+template <bool TrackWhyLive>
+template <class ELFT>
+void MarkLive<TrackWhyLive>::mark() {
   // Mark all reachable sections.
   while (!queue.empty()) {
     InputSectionBase &sec = *queue.pop_back_val();
 
     const RelsOrRelas<ELFT> rels = sec.template relsOrRelas<ELFT>();
     for (const typename ELFT::Rel &rel : rels.rels)
-      resolveReloc(sec, rel, false);
+      resolveReloc<ELFT>(sec, rel, false);
     for (const typename ELFT::Rela &rel : rels.relas)
-      resolveReloc(sec, rel, false);
+      resolveReloc<ELFT>(sec, rel, false);
     for (const typename ELFT::Crel &rel : rels.crels)
-      resolveReloc(sec, rel, false);
+      resolveReloc<ELFT>(sec, rel, false);
 
     for (InputSectionBase *isec : sec.dependentSections)
       enqueue(isec, /*offset=*/0, /*sym=*/nullptr,
@@ -491,8 +487,7 @@ void MarkLive<ELFT, TrackWhyLive>::mark() {
 // We also need to move sections whose names are C identifiers that are referred
 // to from __start_/__stop_ symbols because there will only be one set of
 // symbols for the whole program.
-template <class ELFT, bool TrackWhyLive>
-void MarkLive<ELFT, TrackWhyLive>::moveToMain() {
+template <bool TrackWhyLive> void MarkLive<TrackWhyLive>::moveToMain() {
   for (ELFFileBase *file : ctx.objectFiles)
     for (Symbol *s : file->getSymbols())
       if (auto *d = dyn_cast<Defined>(s))
@@ -507,8 +502,6 @@ void MarkLive<ELFT, TrackWhyLive>::moveToMain() {
         ctx.symtab->find(("__stop_" + sec->name).str()))
       enqueue(sec, /*offset=*/0, /*sym=*/nullptr, /*reason=*/{});
   }
-
-  mark();
 }
 
 // Before calling this function, Live bits are off for all
@@ -530,17 +523,28 @@ template <class ELFT> void elf::markLive(Ctx &ctx) {
     sec->markDead();
 
   // Follow the graph to mark all live sections.
-  for (unsigned i = 1, e = ctx.partitions.size(); i <= e; ++i)
-    if (ctx.arg.whyLive.empty())
-      MarkLive<ELFT, false>(ctx, i).run();
-    else
-      MarkLive<ELFT, true>(ctx, i).run();
+  for (unsigned i = 1, e = ctx.partitions.size(); i <= e; ++i) {
+    if (ctx.arg.whyLive.empty()) {
+      MarkLive<false> marker(ctx, i);
+      marker.seed();
+      marker.template mark<ELFT>();
+    } else {
+      MarkLive<true> marker(ctx, i);
+      marker.seed();
+      marker.template mark<ELFT>();
+      if (i == 1)
+        marker.reportWhyLive();
+    }
+  }
 
   // If we have multiple partitions, some sections need to live in the main
   // partition even if they were allocated to a loadable partition. Move them
   // there now.
-  if (ctx.partitions.size() != 1)
-    MarkLive<ELFT, false>(ctx, 1).moveToMain();
+  if (ctx.partitions.size() != 1) {
+    MarkLive<false> marker(ctx, 1);
+    marker.moveToMain();
+    marker.template mark<ELFT>();
+  }
 
   // Report garbage-collected sections.
   if (ctx.arg.printGcSections.empty())
